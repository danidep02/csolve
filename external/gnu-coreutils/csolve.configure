#!/bin/bash

./config.status
cp -f src/version-csolve.h src/version.h
cd lib
export INCLUDES=-I../../lib
make -f Makefile.csolve
