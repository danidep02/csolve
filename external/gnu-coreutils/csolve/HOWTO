How To Check coreutils Programs With csolve
===========================================

First, configure to build with csolve:

  ./csolve.configure

Then change into the src/ directory and invoke "make.csolve" with the
name of the object file for program you want to build, e.g.,

  ./make.csolve true

(Highly technical note: this compiles true.o, not the final executable
true, to avoid having to build and link in libraries in the lib/
subdirectory.)

Debugging Tip: Checking Programs With A coreutils Build Environment
-------------------------------------------------------------------

If you want to check a program with the same build environment as used
for coreutils programs, you can simply drop an arbitrarily-named file
in the src/ directory and invoke make.csolve with it, as in

  cp tsort.c patsFooBar.c
  ./make.csolve patsFooBar

This is helpful, for example, if you are trying to debug a program
(tsort.c) in this case and you don't want to mutate it in place but
rather make a copy and tweak that.

Version Control
===============

Please modify the programs in place. This will make it easy to look at
the git logs later and see how the source program was modified, if at
all, from the original version.