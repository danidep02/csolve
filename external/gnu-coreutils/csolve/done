Globally, a lot of the examples compare argc and optind. How to
incorporate this reasoning?
aot clear that the spec for quote_n is correct.

* means shape only
+ means with serious work left to be done
;( means blocked on some feature
? means in-progress

true
false
sum
sync
logname
hostid
unlink
+ yes
whoami
link
dirname
hostname
 - double-check; it had weird optind stuff going on and maybe shouldn't have worked
tty
+ groups
+ nproc
  - Uses optarg and getopt. There is a dependency between c, the return value of optarg, the input option structure, and the global variable optarg. Should show that optarg is not null depending on the value of c.
basename    [AB] (with lcc_check_pos uncommented)
* tsort
* mkfifo    [RJ]
libstdbuf   [RJ]   
sleep       [RJ]    -- dependent-addrof issue
;( users    [AB]
 - requires nested arrays
getlimits   [AB]
 - revisit once contextual types are in
 - Need to check csolve_check_pos here
readlink
* setuidgid
 - NUL-terminated array
;( nice
 - NUL-terminated array
+ mknod [PMR]
  - One part of the code checks argv[optind][0] == 'p', then
    another part uses that test; however, we can't propagate that
    info because we can't show that argv[optind] is final, as that
    location is global. (pmr_argchar)
  - We have to fold optind at all function calls because it's in a global
    location, which means we have to manually constant fold for some
    reasoning. (pmr_optind)
  - Complex control flow involving case switches and gotos gets us into
    a situation where the guards imply a fact is true at a block, but no
    guard *dominates* that block, so we have to make a witness variable
    for the fact. (pmr_case_witness)
;( uptime [MK]
  - Contains a non-flat struct in a non-trivial way
? tac-mem [AB]
? nohup [PMR]
? rmdir [PMR]
? runcon [AB]
  - selinux.h is defined with nops that most likely trick csolve into thinking
    Sections of code are not executed. There is certainly more work to be done here.
? chroot [AB]
echo [AB]
? env [MK]
? chgrp [RJ]

---- Inbox/TODO -------------------------------------

chksum
kill
fold
base64
chown

