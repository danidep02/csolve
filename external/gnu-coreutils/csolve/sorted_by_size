       2 false.c
       2 lbracket.c
       2 ls-dir.c
       2 ls-ls.c
       2 ls-vdir.c
       2 uname-arch.c
       2 uname-uname.c
       2 version.c
      20 foo.c
      37 prog-fprintf.c
      74 operand2sig.c
      74 sync.c
      83 true.c
      87 logname.c
      89 hostid.c
      90 unlink.c
      92 yes.c
      94 whoami.c
      95 link.c
     113 find-mount-point.c (lib)
     113 group-list.c (lib)
     114 dirname.c
     117 hostname.c
     125 tty.c
     127 groups.c
     127 nproc.c        
     139 mkfifo.c      
     140 basename.c     
     142 libstdbuf.c        
     151 sleep.c
     152 users.c
     155 printenv.c
     156 env.c
     157 getlimits.c
     165 cp-hash.c
     173 readlink.c
     215 mkdir.c
     217 setuidgid.c
     219 nice.c
     223 tee.c
     232 mknod.c
     241 nohup.c
     253 rmdir.c
     260 uptime.c
     263 tac-pipe.c
     264 runcon.c
     268 chroot.c
     274 echo.c
     288 sum.c
     313 chgrp.c
     315 cksum.c
     315 kill.c
     317 fold.c
     322 base64.c
     326 chown.c
     334 id.c
     346 rm.c
     358 mktemp.c
     359 timeout.c
     369 stdbuf.c
     377 uname.c
     390 pwd.c
     396 truncate.c
     421 shuf.c
     429 pathchk.c
     433 expand.c
     437 touch.c
     443 seq.c
     451 comm.c
     497 mv.c
     508 dircolors.c
     518 su.c
     524 paste.c
     527 chown-core.c
     535 unexpand.c
     551 date.c
     552 chmod.c
     568 uniq.c
     576 chcon.c
     577 factor.c
     584 ln.c
     616 pinky.c
     629 nl.c
     637 remove.c
     675 tac.c
     697 printf.c
     727 tsort.c
     761 md5sum.c
     784 cat.c
     795 wc.c
     842 who.c
     856 test.c
     896 cut.c
     958 df.c
     976 expr.c
     992 du.c
    1011 install.c
    1013 fmt.c
    1064 head.c
    1093 split.c
    1139 join.c
    1175 cp.c
    1217 shred.c
    1471 stat.c
    1491 csplit.c
    1909 stty.c
    1928 od.c
    1949 tr.c
    1954 dd.c
    2203 tail.c
    2222 ptx.c
    2357 copy.c
    2879 pr.c
    4626 sort.c
    4733 ls.c
