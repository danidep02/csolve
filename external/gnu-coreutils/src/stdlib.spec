// /usr/include/stdlib.h:514: Error: No spec for extern function abort. Autogen spec is:

abort ::
  arg       ()
  ret       int(0, 0{1})
  store     []

