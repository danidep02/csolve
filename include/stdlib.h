#ifndef __STDLIB_H__
#define __STDLIB_H__

typedef unsigned int size_t;

typedef size_t ssize_t;

#define NULL   0

extern void *malloc(size_t);

extern long int atol(char * __attribute__ ((array)));

extern void exit(int);

#endif
