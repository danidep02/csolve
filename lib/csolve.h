#ifndef __HAVE_CSOLVE
#define __HAVE_CSOLVE

// Predicate abbreviations
#define PNONNULL       V > 0
#define PNN(p)         ((V != 0) => (p))
#define PVALIDLO       BLOCK_BEGIN([V]) <= V
#define PVALIDHI       V < BLOCK_END([V])

#define PVALIDPTR      && [PNONNULL; PVALIDLO; PVALIDHI]

#define PSTART         V = BLOCK_BEGIN([V])
#define PSIZE(n)       (V + n) = BLOCK_END([V])
#define PSIZE_GE(n)    (V + n) <= BLOCK_END([V])
#define POFFSET(n)     V = (BLOCK_BEGIN([V]) + n)
#define POFFSET_GE(n)  V >= (BLOCK_BEGIN([V]) + n)

#define PEQBLOCK(x)    && [(BLOCK_BEGIN([V]) = BLOCK_BEGIN([x])); (BLOCK_END([V]) = BLOCK_END([x]))]
#define PINTO(x,lo,hi) && [PNONNULL; (PEQBLOCK(x));(POFFSET_GE(lo));(PSIZE_GE((hi + 1)))]
#define PINDEX(x,sz)   && [(BLOCK_BEGIN([x]) <= (x + (sz * V))); ((x + sz + (sz * V)) <= BLOCK_END([x]))]

#ifdef CIL
# define CSOLVE_ATTR(a) __attribute__ ((a))
#else
# define CSOLVE_ATTR(a)
#endif

// Basic macros
// Need to break this into two levels to ensure predicate p is macro expanded
#define SREF(p)            CSOLVE_ATTR (csolve_predicate (#p))
#define REF(p)             SREF(p)
#define NNREF(p)           REF(PNN(p))
#define ARRAY              CSOLVE_ATTR (array)
#define SHAPE_IGNORE_BOUND CSOLVE_ATTR (csolve_ignore_bound)
#define IGNORE_INDEX       CSOLVE_ATTR (csolve_ignore_index)
#define IGNORE_REF_INDEX   IGNORE_INDEX REF(0 = 0) 
#define USE_INDEX          CSOLVE_ATTR (csolve_use_index)
#define ANYREF             CSOLVE_ATTR (csolve_any_ref)
#define ANY                CSOLVE_ATTR (csolve_any_type)
#define NULL               CSOLVE_ATTR (csolve_null_or_ok)
#define VAR(a)             CSOLVE_ATTR (csolve_type_var (#a))

#define FINAL              CSOLVE_ATTR (csolve_final)
#define LOC(l)             CSOLVE_ATTR (csolve_sloc (#l))
#define GLOBAL(l)          CSOLVE_ATTR (csolve_global_loc (#l))
#define OKEXTERN           CSOLVE_ATTR (csolve_extern_ok)
#define CHECK_TYPE         CSOLVE_ATTR (csolve_check_type)
#define INST(l, k)         CSOLVE_ATTR (csolve_inst_sloc (#l, #k))
#define ROOM_FOR(t)        CSOLVE_ATTR (csolve_room_for (sizeof(t)))
#define NNROOM_FOR(t)      CSOLVE_ATTR (csolve_nonnull_room_for (sizeof(t)))

#define HASROOM            CSOLVE_ATTR (csolve_has_room)
#define NNHASROOM          CSOLVE_ATTR (csolve_nonnull_has_room)


#define EFFECT(l, p)       CSOLVE_ATTR (csolve_effect (#l, #p))

// Hack: CIL doesn't allow types as attribute parameters, so we
//       indirect through sizeof.
#define LAYOUT(t)          CSOLVE_ATTR (csolve_layout (sizeof(t)))


// Predicate macros

#define VALIDPTR          REF(PVALIDPTR)
#define NNVALIDPTR        NNREF(PVALIDPTR)
#define NONNULL           REF(PNONNULL)
#define START             REF(PSTART)
#define NNSTART           NNREF(PSTART)
#define SIZE(n)           REF(PSIZE(n))
#define SIZE_GE(n)        REF(PSIZE_GE(n))
#define NNSIZE(n)         NNREF(PSIZE(n))
#define NNSIZE_GE(n)      NNREF(PSIZE_GE(n))
#define OFFSET(n)         REF(POFFSET(n))
#define OFFSET_GE(n)      REF(POFFSET_GE(n))
#define NONNEG            REF(V >= 0)

#define STRINGPTR         ARRAY VALIDPTR
#define NNSTRINGPTR       ARRAY NNVALIDPTR
#define ARRAYSTART        ARRAY START VALIDPTR
#define OK                START VALIDPTR HASROOM
#define NNOK              NNSTART NNVALIDPTR NNHASROOM          
#define SOMEPTR           REF(0 = 0)

#define NULLTERMSTR       REF((VVADDR = (BLOCK_END([VVADDR]) - 1)) => (V = 0))

typedef char NULLTERMSTR * LOC(L) STRINGPTR csolve_string;
typedef char const NULLTERMSTR FINAL * STRINGPTR  csolve_const_string;

// Assumptions

#define CSOLVE_VAR2(base, n) base##n
#define CSOLVE_VAR(base, n)  CSOLVE_VAR2(__csolve__##base, n)
#define CSOLVE_ASSUME(p)     ;int CSOLVE_VAR(assume, __COUNTER__ ) = csolve_assume (p);

// Memory Safety Backdoors

#define UNCHECKED              CSOLVE_ATTR (csolve_unchecked)

#define CSOLVE_UNSAFE_WRITE(p, d) *((typeof (d) * UNCHECKED) p) = d
#define CSOLVE_UNSAFE_READ(p)     *((typeof (p) UNCHECKED) p)

// Built-in functions

extern void csolve_fold_all () OKEXTERN;


extern void validptr (void * VALIDPTR IGNORE_INDEX) OKEXTERN;
extern void csolve_validptr_lo (void * REF(PVALIDLO) IGNORE_INDEX) OKEXTERN;
extern void csolve_validptr_hi (void * REF(PVALIDHI) IGNORE_INDEX) OKEXTERN;

extern int csolve_assert (int REF(V != 0) p) OKEXTERN;

extern void * LOC(L) REF(V = BLOCK_BEGIN([p])) csolve_block_begin (void * LOC(L) IGNORE_REF_INDEX p) OKEXTERN;

extern void * LOC(L) REF(V = BLOCK_END([p])) csolve_block_end (void * LOC(L) IGNORE_REF_INDEX p) OKEXTERN;

extern int REF(b = 1) csolve_assume (int b) OKEXTERN;

extern int nondet () OKEXTERN;

extern int REF(V >= 1) USE_INDEX nondetpos () OKEXTERN;

extern int REF(V >= 0) USE_INDEX nondetnn () OKEXTERN;

extern int REF(V >= l) REF(V < u) REF(V >= 0) USE_INDEX nondetrange (int l, int REF(l < V) u) OKEXTERN;

// Casts
//char * LOC(L) STRINGPTR csolve_check_pos(char * LOC(L) VALIDPTR p) CHECK_TYPE
//{
//  return p;
//}

// Needed for ADPCM -- so put these over there ?
extern int REF(&& [V >= a; V >= b; V >= 0; V <= a + b]) bor (int REF(V >= 0) a, int REF(V >= 0) b) OKEXTERN;

extern int REF(&& [V <= b; V >= 0]) band (int a, int REF(V >= 0) b) OKEXTERN;

extern int REF(&& [V < m; V >= 0; V <= a]) csolve_mod (int REF(V >= 0) IGNORE_INDEX a, int REF(V > 0) IGNORE_INDEX m) OKEXTERN;

extern int REF(&& [ (V = (x * y))
                  ; (V = (y * x))
                  ; (((x * y) + y) = ((x + 1) * y))
                  ; (((x * y) + x) = (x * (y + 1)))
                  ; ((&&[(x > 0); (y > 0)]) => (&&[(V >= x); (V >= y)]))
                  ; ((x = 0) => (V = 0))
                  ; ((y = 0) => (V = 0))
                  ]) USE_INDEX 
  csolve_times (int x, int y) OKEXTERN;

extern int REF((a * c) <= (b * c)) USE_INDEX 
  csolve_axiom_times_mono (int REF(0 <= V) a, int REF(a <= V) b, int REF(0 <= V) c) OKEXTERN;

#endif
