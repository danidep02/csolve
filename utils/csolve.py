#In your .vimrc, add the following:
#map <F4> :python printLiquidType("normal")<CR>
#vmap <F4> :python printLiquidType("visual")<CR>
#pyfile path/to/csolve.py
#to "reparse", do
#:python annot.parseLiquidType()


#python << ENDPYTHON

import vim
import time
import re
import os

class AnnExc(Exception):
  def __init__(self, reason):
    self.reason = reason

class Cannotations:
  def __init__(self):
    self.__vmap_filename = None # last .vmap parsed file
    self.__filename = None 	# last .c parsed file 
    self.__timestamp = None     # last parse action timestamp
    self.__annot = {}
 
  def parse(self, filename):
    self.__filename      = filename 
    self.__vmap_filename = filename + ".vmap" 
    self.__timestamp     = int(time.time())
    try:
      f = open(self.__vmap_filename)
      for l in f:
	es  = [x.strip() for x in l.split("\t")]
	key = (str(es[0]), int(es[2]))
	val = es[3]
	self.__annot[key] = val
      f.close()
    except:
      raise 
      #raise AnnExc("Csolve: Error during __parse") 
 
  def print_keys(self):
    for k in self.__annot:
      print "key = ", k, ", data = ", self.__annot[k]
    return
 
  def get_ssaname(self, varname, line):
    if vim.current.buffer.name == None:
      raise AnnExc("Empty buffer") 
    if vim.current.buffer.name != self.__filename or  \
      os.stat(self.__filename).st_mtime > self.__timestamp:
	filename = self.__filename #vim.current.buffer.name
	self.parse(filename)
	vim.command("set tags=" + filename + ".tags")
    ssaname = varname
    try:
      ssaname = self.__annot[(str(varname), int(line))]
    except:
      pass
    return ssaname
  
  def get_type(self, varname, line):
    ssaname = self.get_ssaname(varname, line)
    return vim.command("tag " + ssaname) 

annot    = Cannotations()	#global annotation object
alphanum = re.compile("^[\w.]$")

def getVarnameVisual(buf):
  (line, col1) = buf.mark("<")
  (_   , col2) = buf.mark(">")
  return buf[line-1][col1 : (col2 + 1)]

def getVarnameNormal(buf, (line, col)):
  left, right = col, col
  while left >= 0 and alphanum.search(buf[line-1][left]):
    left  -= 1
  while right < len(buf[line-1]) and alphanum.search(buf[line-1][right]):
    right += 1
  return buf[line-1][left + 1: right]

def printLiquidType(mode):
  if mode == "visual":
    varname = getVarnameVisual(vim.current.buffer)
  else:
    varname = getVarnameNormal(vim.current.buffer, vim.current.window.cursor)
  try:
    annot.get_type(varname, vim.current.window.cursor[0])
  except:
    print "No annotation for ", varname
    raise

def parseLiquidType():
  annot.parse(vim.current.buffer.name)

def reparseLiquidType(x):
  annot.parse(x)
#ENDPYTHON
