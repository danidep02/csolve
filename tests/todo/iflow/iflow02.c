#include <csolve.h>

#include <stdlib.h>


int *global;

int *foo(int *x) {
  int *y = x;
  return y;
}

void bar(int *x) {
  gloabl = x;
}

void test01(int *x) {
  int *y;
  y = foo(x);
}

void test02(int *x) {
  int *y;
  if (g) {
    foo(x); //does foo need to know that it was called in the context of g?
            //i.e. if it updates a global variable?
  }
}

void test03(int *x) {
  int *y;
  if (g) {
    bar(x); //does bar need to know that it was called in the context of g?
            //i.e. the tags of //global should include the tags of g. But
            //are these even in scope now? The tags of global might be impossible to
            //reason about?
  }
}
