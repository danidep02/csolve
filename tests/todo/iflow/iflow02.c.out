
../tests/todo/iflow/iflow02.c:17:

constraint:
 env  [ fd2_main:{VV : int | [TAGGED0(VV)]}
      ; fd1_main:{VV : int | [TAGGED0(VV)]}
      ; d_main:{VV : ptr(A2) | [&& [ Begin(VV)
                                   ; TAGGED2(VV, fd2_main, fd1_main)
                                   ; TAGGED2(VV, fd1_main, fd2_main)
                                   ; Valid(VV)]]}
      ; UNCHECKED:{VV : func(1, [@(0) ; int]) | []}
      ; TAG:{VV : func(1, [@(0) ; int]) | []}
      ; SKOLEM:{VV : func(0, [int ; int]) | []}
      ; EWRITE:{VV : int | []}
      ; EREAD:{VV : int | []}
      ; DEREF:{VV : func(1, [@(0) ; @(1)]) | []}
      ; BLOCK_END:{VV : func(1, [@(0) ; @(0)]) | []}
      ; BLOCK_BEGIN:{VV : func(1, [@(0) ; @(0)]) | []}] 
 grd true 
 lhs {VV : int | [(VV = fd1_main)]} 
 rhs {VV : int | [(TAG([VV]) = TAG([d_main]))]} 
 id 6 tag [1;1;-1;0] //main 

**********************************************************
******** ERROR ../tests/todo/iflow/iflow02.c:17 Raw   cons_of_call: in 
**********************************************************
